File: backend/src/scraper/cleaner.js
function cleanText(text) {
  if (!text) return '';
  return text.replace(/\s+/g, ' ').trim();
}

function cleanPhone(phone) {
  if (!phone) return '';
  return phone.replace(/[^\d+\s\-\(\)]/g, '').trim();
}

function cleanEmail(email) {
  if (!email) return '';
  return email.trim().toLowerCase();
}

function cleanUrl(url) {
  if (!url) return '';
  let cleaned = url.trim();
  if (cleaned && !cleaned.startsWith('http://') && !cleaned.startsWith('https://')) {
    cleaned = 'https://' + cleaned;
  }
  return cleaned.replace(/\/+$/, '');
}

function isValidPhone(phone) {
  if (!phone) return false;
  const digits = phone.replace(/[^\d]/g, '');
  return digits.length >= 8 && digits.length <= 15;
}

function isValidEmail(email) {
  if (!email) return false;
  const pattern = /^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$/;
  return pattern.test(email.trim());
}

function deduplicatePhones(phones) {
  const seen = new Set();
  const unique = [];
  for (const p of phones) {
    const raw = p.normalized || p.number || '';
    const norm = raw.replace(/[^\d]/g, '');
    if (!seen.has(norm) && norm.length >= 8) {
      seen.add(norm);
      unique.push(p);
    }
  }
  return unique;
}

function deduplicateEmails(emails) {
  const seen = new Set();
  const unique = [];
  for (const e of emails) {
    const norm = (e.normalized || e.email || '').toLowerCase();
    if (!seen.has(norm)) {
      seen.add(norm);
      unique.push(e);
    }
  }
  return unique;
}

function deduplicateSocialLinks(links) {
  const seen = new Set();
  const unique = [];
  for (const link of links) {
    const url = link.url || '';
    if (!seen.has(url)) {
      seen.add(url);
      unique.push(link);
    }
  }
  return unique;
}

function emptyToNull(val) {
  if (val === null || val === undefined || (typeof val === 'string' && val.trim() === '')) {
    return null;
  }
  return val;
}

module.exports = {
  cleanText,
  cleanPhone,
  cleanEmail,
  cleanUrl,
  isValidPhone,
  isValidEmail,
  deduplicatePhones,
  deduplicateEmails,
  deduplicateSocialLinks,
  emptyToNull,
};

File: backend/app/scraper/cleaner.py
import re
from typing import List, Dict, Optional


def clean_text(text: str) -> str:
    if not text:
        return ""
    return re.sub(r"\s+", " ", text).strip()


def clean_phone(phone: str) -> str:
    return re.sub(r"[^\d+\s\-\(\)]", "", phone).strip()


def clean_email(email: str) -> str:
    return email.strip().lower()


def clean_url(url: str) -> str:
    url = url.strip()
    if url and not url.startswith(("http://", "https://")):
        url = "https://" + url
    return url.rstrip("/")


def is_valid_phone(phone: str) -> bool:
    if not phone:
        return False
    digits = re.sub(r"[^\d]", "", phone)
    # Must have 10 digits (Indian mobile) or more for landlines
    return 8 <= len(digits) <= 15


def is_valid_email(email: str) -> bool:
    if not email:
        return False
    pattern = re.compile(r"^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$")
    return bool(pattern.match(email.strip()))


def deduplicate_phones(phones: List[Dict]) -> List[Dict]:
    seen = set()
    unique = []
    for p in phones:
        norm = re.sub(r"[^\d]", "", p.get("normalized") or p.get("number", ""))
        if norm not in seen and len(norm) >= 8:
            seen.add(norm)
            unique.append(p)
    return unique


def deduplicate_emails(emails: List[Dict]) -> List[Dict]:
    seen = set()
    unique = []
    for e in emails:
        norm = (e.get("normalized") or e.get("email", "")).lower()
        if norm not in seen:
            seen.add(norm)
            unique.append(e)
    return unique


def deduplicate_social_links(links: List[Dict]) -> List[Dict]:
    seen = set()
    unique = []
    for link in links:
        url = link.get("url", "")
        if url not in seen:
            seen.add(url)
            unique.append(link)
    return unique


def empty_to_none(value: Optional[str]) -> Optional[str]:
    if value is None or value.strip() == "":
        return None
    return value
